BUG: `scipy/integrate/__quadpack.c` — `dqk15w` forward declaration uses K&R `double()`, fails to compile under C23

### Describe your issue.

On commit `d4a7a2463e1eb2825848a61f6913013fd1f5c21b` `scipy/integrate/__quadpack.c` does not compile with C23. The forward declaration of the internal `dqk15w` routine declares its weight-function parameter with the empty-parentheses K&R form `double()`. 

In c23 an empty parameter list means `(void)` (not "unspecified arguments" as in C17 and earlier). So the forward declaration's second parameter becomes `double (*)(void)`, which conflicts with
the function's own definition (which uses `quadpack_w_func`) and rejects the weight functions passed at every
call site.

The file already defines the proper weight-function type and uses it in `dqk15w`'s definition. Only the
forward declaration seems to be out of sync.

Under C17 the K&R `double()` was compatible with `quadpack_w_func` ("unspecified args"), so this compiled. Under C23, `double()` is `double(void)`, so the declaration and definition disagree.


The fix is to ,ake the forward declaration use the same `quadpack_w_func` type as the definition. This is C17 and 
C23-correct.

Current sed patch: 
```sh
sed -i 's/,double(),/,quadpack_w_func,/' scipy/integrate/__quadpack.c
```

### Reproducing Code Example

```bash
# Compile from sources with C23.

pip3 install . --compile --no-dependencies \ 
  -Csetup-args="-Dblas=openblas" \
  -Csetup-args="-Dlapack=openblas" \
  -Csetup-args="-Duse-pythran=true" \
  -Csetup-args=-Db_colorout=always \
  -Csetup-args=-Dc_std=gnu23 \
  -Csetup-args=-Dcpp_std=gnu++17 # Must downgrade to c++17 which is a separate issue
```

### Error message

```shell
scipy/integrate/__quadpack.c:3255:21: error: incompatible function pointer types passing
      'double (const double, const double, const double, const double, const double, const int)'
      to parameter of type 'double (*)(void)' [-Wincompatible-function-pointer-types]
scipy/integrate/__quadpack.c:3447:21: error: incompatible function pointer types ...
scipy/integrate/__quadpack.c:4102:17: error: incompatible function pointer types ...
scipy/integrate/__quadpack.c:4874:1: error: conflicting types for 'dqk15w'
```

### SciPy/NumPy/Python version and system information

```shell
Commit `d4a7a2463e1eb2825848a61f6913013fd1f5c21b`
Clang20
```

--- a/scipy/integrate/__quadpack.c
+++ b/scipy/integrate/__quadpack.c
@@ -170,7 +170,7 @@ static void dqcheb(const double*,double*,double*,double*);
 static void dqelg(int*,double*,double*,double*,double*,int*);
 static void dqk15(double(*)(double*),const double,const double,double*,double*,double*,double*);
 static void dqk15i(double(*)(double*),const double,const int,const double,const double,double*,double*,double*,double*);
-static void dqk15w(double(*)(double*),double(),const double,const double,
+static void dqk15w(double(*)(double*),quadpack_w_func,const double,const double,
                    const double,const double,const int,const double,const double,
                    double*,double*,double*,double*);
 static void dqk21(double(*)(double*),const double,const double,double*,double*,double*,double*);

